crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,376 @@
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// # fn sys_d() {}
+/// # fn sys_e() {}
+/// # fn sys_f() {}
+/// let graph = SystemGraph::new();
+/// graph.root(sys_a.system())
+///      .fork((sys_b.system(), sys_c.system(), sys_d.system()))
+///      .join(sys_e.system())
+///      .then(sys_f.system());
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+///
+/// # Cloning
+/// This type is backed by a Rc, so cloning it will still point to the same logical
+/// underlying graph.
+///
+/// [fork]: crate::schedule::SystemGraphNode::join
+/// [join]: crate::schedule::SystemJoin::join
+#[derive(Clone)]
+pub struct SystemGraph {
+    id: u32,
+    nodes: Rc<RefCell<HashMap<NodeId, SystemDescriptor>>>,
+}
+
+impl SystemGraph {
+    pub fn new() -> Self {
+        Self {
+            id: NEXT_GRAPH_ID.fetch_add(1, Ordering::Relaxed),
+            nodes: Default::default(),
+        }
+    }
+
+    /// Creates a root graph node without any dependencies. A graph can have multiple distinct
+    /// root nodes.
+    pub fn root(&self, system: impl Into<SystemDescriptor>) -> SystemGraphNode {
+        self.create_node(system.into())
+    }
+
+    /// Checks if two graphs instances point to the same logical underlying graph.
+    pub fn is_same_graph(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+
+    fn create_node(&self, mut system: SystemDescriptor) -> SystemGraphNode {
+        let mut nodes = self.nodes.borrow_mut();
+        assert!(
+            nodes.len() <= u32::MAX as usize,
+            "Cannot add more than {} systems to a SystemGraph",
+            u32::MAX
+        );
+        let id = NodeId {
+            graph_id: self.id,
+            node_id: nodes.len() as u32,
+        };
+        match &mut system {
+            SystemDescriptor::Parallel(descriptor) => descriptor.labels.push(id.dyn_clone()),
+            SystemDescriptor::Exclusive(descriptor) => descriptor.labels.push(id.dyn_clone()),
+        }
+        nodes.insert(id, system);
+        SystemGraphNode {
+            id,
+            graph: self.clone(),
+        }
+    }
+
+    fn add_dependency(&self, src: NodeId, dst: NodeId) {
+        if let Some(system) = self.nodes.borrow_mut().get_mut(&dst) {
+            match system {
+                SystemDescriptor::Parallel(descriptor) => descriptor.after.push(src.dyn_clone()),
+                SystemDescriptor::Exclusive(descriptor) => descriptor.after.push(src.dyn_clone()),
+            }
+        } else {
+            panic!(
+                "Attempted to add dependency for {:?}, which doesn't exist.",
+                dst
+            );
+        }
+    }
+}
+
+/// A draining conversion to [SystemSet]. All other clones of the same graph will be empty
+/// afterwards.
+///
+/// [SystemSet]: crate::schedule::SystemSet
+impl From<SystemGraph> for SystemSet {
```

Comment:
Also, I forget what the API for `add_system_set` is, however we can probably update it to
```rust
fn add_system_set(set: impl Into<SystemSet>)
```
That way we don't have to call `graph.into()` when passing them as this parameter :)